python/dgl/dataloading/dataloader.py

```diff
@@ -1,900 +1,602 @@
+            self._tensor_dataset = indices_shared
+            self._device = indices_shared.device
+
+    def shuffle(self):
+        """Shuffles the dataset."""
+        # Only rank 0 does the actual shuffling.  The other ranks wait for it.
+        if self.rank == 0:
+            self._tensor_dataset[:self.num_indices] = self._tensor_dataset[
+                torch.randperm(self.num_indices, device=self._device)]
+            if not self.drop_last:
+                # pad extra
+                self._tensor_dataset[self.num_indices:] = \
+                    self._tensor_dataset[:self.total_size - self.num_indices]
+        dist.barrier()
+
+    def __iter__(self):
+        start = self.num_samples * self.rank
+        end = self.num_samples * (self.rank + 1)
+        dataset = _divide_by_worker(self._tensor_dataset[start:end])
+        return _TensorizedDatasetIter(
+            dataset, self.batch_size, self.drop_last, self._mapping_keys)
+
+    def __len__(self):
+        return (self.num_samples + (0 if self.drop_last else (self.batch_size - 1))) // \
+            self.batch_size
+
+
+def _prefetch_update_feats(feats, frames, types, get_storage_func, id_name, device, pin_memory):
+    for tid, frame in enumerate(frames):
+        type_ = types[tid]
+        default_id = frame.get(id_name, None)
+        for key in frame.keys():
+            column = frame[key]
+            if isinstance(column, LazyFeature):
+                parent_key = column.name or key
+                if column.id_ is None and default_id is None:
+                    raise DGLError(
+                        'Found a LazyFeature with no ID specified, '
+                        'and the graph does not have dgl.NID or dgl.EID columns')
+                feats[tid, key] = get_storage_func(parent_key, type_).fetch(
+                    column.id_ or default_id, device, pin_memory)
+
+
+# This class exists to avoid recursion into the feature dictionary returned by the
+# prefetcher when calling recursive_apply().
+class _PrefetchedGraphFeatures(object):
+    __slots__ = ['node_feats', 'edge_feats']
+    def __init__(self, node_feats, edge_feats):
+        self.node_feats = node_feats
+        self.edge_feats = edge_feats
+
+
+def _prefetch_for_subgraph(subg, dataloader):
+    node_feats, edge_feats = {}, {}
+    _prefetch_update_feats(
+        node_feats, subg._node_frames, subg.ntypes, dataloader.graph.get_node_storage,
+        NID, dataloader.device, dataloader.pin_memory)
+    _prefetch_update_feats(
+        edge_feats, subg._edge_frames, subg.canonical_etypes, dataloader.graph.get_edge_storage,
+        EID, dataloader.device, dataloader.pin_memory)
+    return _PrefetchedGraphFeatures(node_feats, edge_feats)
+
+
+def _prefetch_for(item, dataloader):
+    if isinstance(item, DGLHeteroGraph):
+        return _prefetch_for_subgraph(item, dataloader)
+    elif isinstance(item, LazyFeature):
+        return dataloader.other_storages[item.name].fetch(
```

Comment:
Does `dataloader` always have an `other_storages` attribute?

Reply:
Not for `GraphDataLoader`.